src/bbs.rs

```diff
@@ -0,0 +1,287 @@
+                let mut cursor = std::io::Cursor::new(arr);
+                let value = Fr::deserialize(&mut cursor, true).map_err(|_| serde::de::Error::invalid_value(serde::de::Unexpected::Bytes(&arr), &self))?;
+                Ok(BlsSecretKey(value))
+            }
+        }
+
+        deserializer.deserialize_tuple(32, SecretKeyVisitor)
+    }
+
+}
+
+/// Generate a blinded BLS key pair where secret key `x` and blinding factor `r` in Fr
+/// and public key `W` = `G2` * `x` * `Blinding_G2` * `r`
+pub fn bls_generate_blinded_g2_key() -> BlsKeyPair<G2> {
+    bls_generate_keypair(None, Some(BLINDING_G2))
+}
+
+/// Generate a blinded BLS key pair where secret key `x` and blinding factor `r` in Fr
+/// and public key `W` = `G1` * `x` * `Blinding_G1` * `r`
+pub fn bls_generate_blinded_g1_key() -> BlsKeyPair<G1> {
+    bls_generate_keypair(None, Some(BLINDING_G1))
+}
+
+/// Generate a BLS key pair where secret key `x` in Fr
+/// and public key `W` = `G1` * `x`
+pub fn bls_generate_g2_key() -> BlsKeyPair<G2> {
+    bls_generate_keypair(None, None)
+}
+
+/// Generate a BLS key pair where secret key `x` in Fr
+/// and public key `W` = `G1` * `x`
+pub fn bls_generate_g1_key() -> BlsKeyPair<G1> {
+    bls_generate_keypair(None, None)
+}
+
+/// A BLS public key where the public key can be either in G1 or G2
+#[derive(Copy, Clone, Debug)]
+pub struct BlsPublicKey<G: CurveProjective<Engine = Bls12, Scalar = Fr>>(pub G);
+
+impl<G: CurveProjective<Engine = Bls12, Scalar = Fr>> Zeroize for BlsPublicKey<G> {
+    fn zeroize(&mut self) {
+        let tv = self.0;
+        self.0.sub_assign(&tv);
+    }
+}
+
+impl BlsPublicKey<G2> {
+    pub fn to_bbs_public_key(&self, message_count: usize) -> PublicKey {
+        let dpk = DeterministicPublicKey::from(self.0);
+        dpk.to_public_key(message_count).unwrap()
+    }
+}
+
+/// A BLS keypair where the public key can be in either G1 or G2
+pub struct BlsKeyPair<G: CurveProjective<Engine = Bls12, Scalar = Fr> + SerDes> {
+    pub secret_key: BlsSecretKey,
+    pub public_key: BlsPublicKey<G>,
+    pub blinder: Option<Fr>,
+}
+
+impl<G: CurveProjective<Engine = Bls12, Scalar = Fr> + SerDes> BlsKeyPair<G> {
+    pub fn new(seed: Option<&[u8]>, blinder: Option<&[u8]>) -> Self {
+        bls_generate_keypair(seed, blinder)
+    }
+}
+
+impl From<BlsSecretKey> for SecretKey {
+    fn from(k: BlsSecretKey) -> Self {
+        SecretKey::from(k.0)
+    }
+}
+
+impl BlsSecretKey {
+    pub fn to_bbs_public_key(&self, message_count: usize) -> PublicKey {
+        let mut g2 = G2::one();
+        g2.mul_assign(self.0);
+        let dpk = DeterministicPublicKey::from(g2);
+        dpk.to_public_key(message_count).unwrap()
+    }
+}
+
+fn bls_generate_keypair<G: CurveProjective<Engine = Bls12, Scalar = Fr> + SerDes>(
+    seed: Option<&[u8]>,
+    blinder: Option<&[u8]>,
+) -> BlsKeyPair<G> {
+    let mut passed_seed = seed.is_some();
+    let seed = match seed {
+        Some(arg) => {
+            passed_seed = true;
```

Comment:
```suggestion
```
This line seems unnecessary, as the variable is already set earlier in the function with the same value.

Alternatively, assignment could be done in the other case:
```diff
@@ -223,13 +223,14 @@ fn bls_generate_keypair<G: CurveProjective<Engine = Bls12, Scalar = Fr> + SerDes
     seed: Option<&[u8]>,
     blinder: Option<&[u8]>,
 ) -> BlsKeyPair<G> {
-    let mut passed_seed = seed.is_some();
+    let mut passed_seed;
     let seed = match seed {
         Some(arg) => {
             passed_seed = true;
             arg.to_vec()
         }
         None => {
+            passed_seed = false;
             let mut rng = thread_rng();
             let mut seed_data = vec![0u8, 32];
             rng.fill_bytes(seed_data.as_mut_slice());

```